### PR TITLE
Reconcile control plane in vsphere

### PR DESCRIPTION
**Pull request by ddjjia**
*Issue #, if available:*
Issue #3917 

*Description of changes:*
Implement the ProviderClusterReconciler interface: ControlPlane reconciliation

*Testing (if applicable):*
unit test

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/providers/vsphere/reconciler/reconciler_test.go [hunk 1]
```diff
@@ -149,6 +149,17 @@ func TestReconcileCNIErrorClientRegistry(t *testing.T) {
...
 
+func TestReconcilerReconcileControlPlaneSuccess(t *testing.T) {
+	tt := newReconcilerTest(t)
+	tt.createAllObjs()
+
+	result, err := tt.reconciler().ReconcileControlPlane(tt.ctx, test.NewNullLogger(), tt.buildSpec())
+
+	tt.Expect(err).NotTo(HaveOccurred())
+	tt.Expect(tt.cluster.Status.FailureMessage).To(BeZero())
+	tt.Expect(result).To(Equal(controller.Result{}))
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler_test.go:160]: I would be great if we could check that the capi objects for workers have been created :)

Comment (ddjjia) [reply pkg/providers/vsphere/reconciler/reconciler_test.go:160]: You mean check if the control plane objects are created during the test right? How do you capture a variable created in runtime without return it

Comment (g-gaston) [reply pkg/providers/vsphere/reconciler/reconciler_test.go:160]: Yeah exactly that
Ahh great question
You can use the client.Client to read it from the cluster :)

File: pkg/providers/vsphere/reconciler/reconciler_test.go [hunk 2]
```diff
@@ -149,6 +150,26 @@ func TestReconcileCNIErrorClientRegistry(t *testing.T) {
...
+
+	obj := &addonsv1.ClusterResourceSet{ObjectMeta: metav1.ObjectMeta{
+		Name:      "workload-cluster-crs-0",
+		Namespace: "eksa-system",
+	}}
+	key := client.ObjectKeyFromObject(obj)
+	tt.Eventually(func() error {
+		return tt.client.Get(tt.ctx, key, obj)
+	}).Should(Succeed(), "\"object %s should eventually exist", obj.GetName())
```
Comment (g-gaston) [inline pkg/providers/vsphere/reconciler/reconciler_test.go:170]: Can we do the same thing for the other CP objects? at least the obvious ones: capi cluster, kcp and vsphere machine template
I'm approving the PR, I think you can add it as a follow up PR :)

**Review by g-gaston — APPROVED** [review]

**Review by taneyland — APPROVED** [review]

**Pull request merged**